content/courses/Ada_For_The_C_Embedded_Developer/chapters/07_Reusability.rst

```diff
@@ -497,10 +497,270 @@ Object orientation
 Pointer to subprograms
 ~~~~~~~~~~~~~~~~~~~~~~
 
-.. todo::
+Pointers to subprograms allow us to dynamically select an appropriate
+subprogram at runtime. This selection might be triggered by an external
+event, or simply by the user. This can be useful when multiple versions of a
+routine exist, and the decision about which one to use cannot be made at
+compilation time.
 
-    Complete section!
+This is an example on how to declare and use pointers to functions in C:
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Reusability.Selecting_Subprogram_C
+
+    !main.c
+    #include <stdio.h>
+    #include <stdlib.h>
+
+    void show_msg_v1 (char *msg)
+    {
+        printf("Using version #1: %s\n", msg);
+    }
+
+    void show_msg_v2 (char *msg)
+    {
+        printf("Using version #2:\n %s\n", msg);
+    }
+
+    int main()
+    {
+        int selection = 1;
+        void (*current_show_msg) (char *);
+
+        switch (selection)
+        {
+            case 1:  current_show_msg = &show_msg_v1;    break;
+            case 2:  current_show_msg = &show_msg_v2;    break;
+            default: current_show_msg = NULL;            break;
+        }
+
+        if (current_show_msg != NULL)
+        {
+            current_show_msg ("Hello there!");
+        }
+        else
+        {
+            printf("ERROR: no version of show_msg() selected!\n");
+        }
+    }
+
+The example above contains two versions of the :c:`show_msg()` function:
+:c:`show_msg_v1()` and :c:`show_msg_v2()`. The function is selected depending
+on the value of :c:`selection`, which initializes the function pointer
+:c:`current_show_msg`. If there's no corresponding value, :c:`current_show_msg`
+is set to :c:`null` |mdash| alternatively, we could have selected a default
+version of :c:`show_msg()` function. By calling
+:c:`current_show_msg ("Hello there!")`, we're calling the function that
+:c:`current_show_msg` is pointing to.
+
+This is the corresponding implementation in Ada:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Reusability.Selecting_Subprogram_Ada
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Main is
```

Comment:
I think we should consider a policy of typically not using "Main" as the name of the main subprogram because C/C++ people may conclude that it is required, as in their languages. What do you think?

Reply:
Very good idea. I'll make sure to avoid _Main_ in the Ada examples.